dopamine/discrete_domains/run_experiment.py

```diff
@@ -540,13 +545,23 @@ def _save_tensorboard_summaries(self, iteration, num_episodes,
     self._summary_writer.add_summary(summary, iteration)
 
 
+def async_method(method):
```

Comment:
Do you prefer this function in this module over the threading utils module?

Reply:
For now yes because we might need to make it specific to this module.

Though if we can we can move it later on. Does that work for you?

Reply:
It's a little hard to comment about what should be done with this function without picturing your future plans for it.  As written, I believe these two things about it:

* Its functionality should be inlined into `_run_experiment_loop`.  It does a non-obvious thing that's invisible to the caller and makes reading `_run_experiment_loop` harder to understand.
* The function's generality make it a better candidate for `utils.threading_utils`.

I trust your picture of what you've got in mind for it, and trust that in light of those plans it's best defined in its own function and in this class.  I just want to bring those above concerns to your attention so you can explicitly approve it.

What sort of changes to this function are you imagining?  (I intend this question as an offer to help you think about the implementation -- we could have it over GVC or not at all, as per your interest).

Reply:
The next change that I wanted to make is to have it handle the semaphore calls.

I might also need it in a later step to decouple training from experience generation.

Moved to `threading_utils`.

```
It does a non-obvious thing that's invisible to the caller and makes reading _run_experiment_loop harder to understand.
```
Yes that's what I liked about the implementation: the fact that it looks as if there was nothing asynchronous :)

Reply:
>> Its functionality should be inlined into _run_experiment_loop. It does a non-obvious thing that's invisible to the caller and makes reading _run_experiment_loop harder to understand.

> Yes that's what I liked about the implementation: the fact that it looks as if there was nothing asynchronous :)

For me, this kind of magic is generally a misfeature.  I want the behavior of code I read to be as apparent as possible, even when that clarity comes with extra verbosity.

Reply:
Writing this functionality into a decorator instead of inlining it into `_run_experiment_loop` makes me anxious.  But this is your implementation, and you've got a better picture of it than I do.  It is also consistent with the approach taken in @local_attributes.

Whatever you choose to do here, LGTM.

Reply:
Moved inline to `_run_iterations`.

You are right. I was anticipating the need for using it when decoupling training from episode generation, but it's something to avoid.